posts/2022-10-06-experimenting-with-voting-schemes.adoc

```diff
@@ -0,0 +1,400 @@
+
+==== Example
+
+In the following example we examine the encryption, combination and decryption of Alice's and Bob's votes using ElGamal encryption. This illustrates how the ballots are encrypted in practice and should give a more concrete intuition of how the votes stay private.
+
+.Details of Alice's and Bob's encrypted ballots
+[%collapsible]
+====
+
+* Alice wants to vote for the first candidate and Bob for the second: `v~a~ = (1, 0, 0)` and `v~b~ = (0, 1, 0)`
+* let `q = 53`, `p = 2q + 1 = 107` and `g = 75`: `+g+` generates a subgroup of `Z/pZ` of order `q`
+* (note that we have not mentioned `q` earlier for simplicity, but it does not change the rest of this example)
+* let the key pair be `x = 18` and `h = g^x^ = 35 mod p`
+* the encrypted ballots can be calculated as follows: 
+** choose a random value, say `r = 2`
+** encryption of the first element of Alice's vote: `b~a1~ = (g^r^, h^r^ * g^m^) = (61, 48 * 75) = (61, 69)`
+** note that the vote is projected in `Z/pZ` before being encrypted with `g^m^`
+** similarly for the first element of Bob's vote with `r = 3`: `b~b1~ = (81, 75 * 1) = (81, 75)`
+* the encrypted sum for the first candidate is: `c~1~ = b~a1~ * b~b1~ = (61 * 81, 69 * 75) = (19, 39)`
+* and once decrypted: `s~1~ = 39 * (19^x^)^-1^ = 39 * 69^-1^ = 39 * 76 = 75 = g^1^`
+* the sum of the votes for the first candidate is indeed `1`
+====
+
+=== Zero-knowledge proofs [[pok]]
+
+Now that we know how the votes are kept private, the verifiability chain remains to be completed with proofs of proper encryption and decryption.
+
+In general, a https://en.m.wikipedia.org/wiki/Zero-knowledge_proof[zero-knowledge proof] is a protocol where a prover P attempts to convince a verifier V that she holds a piece of information without revealing anything about it. An typical example is https://www.zkdocs.com/docs/zkdocs/zero-knowledge-protocols/schnorr/[Schorr's protocol], where P proves to V that she knows the discrete logarithm `x` such that `h = g^x^` (in modular arithmetic).
+
+.Schorr's protocol
+----
++-----------------+                +-----------------+
+|     Prover      |                |    Verifier     |
+|                 |                |                 |
+| choose random r |                |                 |
+|     u = g^r     |       u        |                 |
+|                 | -------------> |                 |
+|                 |                | choose random c |
+|                 |       c        |                 |
+|                 | <------------- |                 |
+|  z = r + x * c  |                |                 |
+|                 |       z        |                 |
+|                 | -------------> |                 |
+|                 |                |    check if     |
+|                 |                |  g^z = u * h^c  |
++-----------------+                +-----------------+
```

Comment:
Maybe it would be good to point out what is private and public information in this scheme?